src/stateManager/ValidationService.ts

```diff
@@ -1,384 +1,494 @@
+            );
+        } else {
+            previousStateSnapshot = previousblockOrSnapshot.stateSnapshot;
+            previousTimestamp = previousStateSnapshot!.timestamp;
         }
 
-        if (!allowedParticipantsSet.has(confirmer)) {
-            return dispute(AgreementFlag.INVALID_SIGNATURE);
+        const isValidTimestamp =
+            block.timestamp >= previousTimestamp &&
+            block.timestamp <= previousTimestamp + this.timeConfig.p2pTime;
+
+        if (!isValidTimestamp) {
+            if (
+                block.height <= 0 || // genesis block
+                previousBlock === undefined || // no previous block == genesis block
+                previousBlock.onChainTimestamp !== undefined
+            ) {
+                // already has best timestamp
+                // no point in trying to update the timestamp => fraud proof
+                this.fraudProofService.createInvalidTimestampProof(
+                    block.coordinates,
+                    signedBlock
+                );
+                return {
+                    shouldDisconnect: true,
+                    action: BlockValidationAction.DISPUTE
+                };
+            }
+
+            // try to fetch later timestamp from on-chain data
+            const onChainTimestamp = await this.fetchOnChainTimestamp(
+                previousBlock,
+                channelId
+            );
+
+            // Early return: Couldn't fetch or not better than current
+            if (!onChainTimestamp || onChainTimestamp <= previousTimestamp) {
+                this.fraudProofService.createInvalidTimestampProof(
+                    block.coordinates,
+                    signedBlock
+                );
+                return {
+                    shouldDisconnect: true,
+                    action: BlockValidationAction.DISPUTE
+                };
+            }
+
+            // Update the previous block with the on-chain timestamp
+            previousBlock.onChainTimestamp = onChainTimestamp;
+            this.storage.blocks.setOnChainTimestamp(
+                previousBlock.forkId,
+                previousBlock.height,
+                onChainTimestamp
+            );
+
+            // Re-validate with updated timestamp
+            const isValidTimestamp =
+                block.timestamp >= onChainTimestamp &&
+                block.timestamp <= onChainTimestamp + this.timeConfig.p2pTime;
+
+            if (!isValidTimestamp) {
+                this.fraudProofService.createInvalidTimestampProof(
+                    block.coordinates,
+                    signedBlock
+                );
+                return {
+                    shouldDisconnect: true,
+                    action: BlockValidationAction.DISPUTE
+                };
+            }
+            // If valid, continue with rest of validation
+
+            if (
+                block.onChainTimestamp === undefined ||
+                block.onChainTimestamp <= block.timestamp
```

Comment:
this `if` should be on the outside of the initial/first `if(!isValidTimestamp)` - take a look at the diagram in the diagram branch under the blockConfirmation pipeline - it has TimeLogic

Reply:
this is enough for this PR
will be done in follow

https://trello.com/c/gI7XhT3M